Spigot-Server-Patches/0407-Fixed-MC-156852.patch

```diff
@@ -0,0 +1,28 @@
+From 73127de4ba34855c821c62e59dd18f47b89bcd04 Mon Sep 17 00:00:00 2001
+From: TheGreatKetchup <TheGreatKetchup@users.noreply.github.com>
+Date: Thu, 1 Aug 2019 21:24:30 -0400
+Subject: [PATCH] Fixed MC-156852
+
+This corrects the 1.14.4 of "phantom" blocks that the client thinks are
+deleted but the server does not.
+
+It uses the same solution that fixed the glitch that caused the same
+issue in 1.8-1.12.
+
+Originally solved by Gnembon on MC-5694 at bugs.mojang.com
+
+diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
+index e5e9de542..3e338ad45 100644
+--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
++++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
+@@ -218,6 +218,7 @@ public class PlayerInteractManager {
+                     int j = (int) (f * 10.0F);
+ 
+                     this.world.a(this.player.getId(), blockposition, j);
++                    this.world.notify(blockposition, iblockdata, iblockdata, 0); // Paper - fixes MC-156852
```

Comment:
I was looking at this and was concerned that this seemed more invasive of a method than I recall, it seems like just sending the block back to the player should be food enough?
`this.player.playerConnection.sendPacket(new PacketPlayOutBlockChange(this.world, blockposition));`

Reply:
I will investigate that further

Reply:
From looking further, it seems like world.notify() queues the block to be updated with `PacketPlayOutMultiBlockChange()`. (Really it marks that block in this chunk for "ticking".)

While ```java
this.player.playerConnection.sendPacket(new PacketPlayOutBlockChange(this.world, blockposition));```

Obviously immediately sends a packet to the client.

`world.notify()` may be more "invasive", it is not necessarily slower as it will combine packets.  However it seems very difficult to actually profile which is faster.

That said, from my testing on just timing the two function calls, directly sending the packet was ~.04 ms faster than using `world.notify()`. So my best guest is the speed does not matter, compared to network latency of at least 5 ms.

However in my testing, I did actually find that by simply deleting the `new PacketPlayOutBlockBreak` the issue is also resolved.

So does anyone have any thoughts on just deleting `new PacketPlayOutBlockBreak`.
Frankly, it would be the best solution if by eliminating this call the problem could be solved, but my gut says this must break something. Can someone prove me wrong?

Reply:
It's not about what's faster or slower, it's about the unintended side effects. Just sending a packet directly is much safer.

Reply:
The issue here is ultimately the state desync between the client and the server, which in most cases is only a single block, no other clients should be affected by this state desync as they won't have been told that the block actually broke (whereas the client breaking the block makes its own assumption over what happened)

using notify means that we're going to mark the chunk as having a dirty block, which will then be sent to all clients that can see the chunk, both of which actions are redundant:
1) The chunk is not dirty (It's not been changed, just the client is making assumptions, meaning that we're inducing extra work to update this and run additional checks which aren't needed)
2) The only client out of sync is the client breaking the block

Reply:
Agreed. Thank you very much for the detailed breakdown and suggestions, I will make the changes.